crates/emitter/src/control_structures.rs

```diff
@@ -0,0 +1,220 @@
+use super::emitter::BytecodeOffset;
+use crate::ast_emitter::AstEmitter;
+use crate::emitter::InstructionWriter;
+
+// Control structures
+
+#[derive(Debug)]
+pub enum JumpKind {
+    Coalesce,
+    LogicalAnd,
+    LogicalOr,
+    IfEq,
+    Goto,
+}
+
+trait Jump {
+    fn jump_kind(&mut self) -> &JumpKind {
+        &JumpKind::Goto
+    }
+
+    fn should_fallthrough(&mut self) -> bool {
+        // a fallthrough occurs if the jump is a conditional jump and if the
+        // condition doesn't met, the execution goes to the next opcode
+        // instead of the target of the jump.
+        match self.jump_kind() {
+            JumpKind::Goto { .. } => false,
+            _ => true,
```

Comment:
now I wonder if it's better explicitly enumerating other cases,
so that whenever new jump kind is added, it will cause compile error and let the author to decide which one the new jump kind should belong to,
instead of silently make it fallthrough.